### ISSUE TITLE
Review

### ISSUE THREAD

**Issue by pamtbaau** [bitstarr/grav-plugin-markdown-details#8]
I've taken a brief look at the plugin and would like to share my findings. Hope you may find it useful.

- Since the advent of Grav 1.7.0 things have change a bit. You might want to take a look at the [upgrade docs](https://learn.getgrav.org/17/advanced/grav-development/grav-17-upgrade-guide#developers) and especially the section for developers.
An easy upgrade approach could be to create a new plugin template using `$ bin/plugin devtools new-plugin` and copy your existing code into the right places.
- The events `onMarkdownInitialized` and `onTwigSiteVariables` are being handled on each and every request, also when derived from Admin. Might be a good idea to move the event registrations into `onPluginsInitialized` and enable them only when request is not from Admin.
- There is a dedicated event `onAssetsInitialized` where assets can be added.
  > onAssetsInitialized
  The event indicates the assets manager has been initialized and is ready for assets to be added and managed.
- Usually, assets like `js` and `css` are stored in their respective folders `/js` and `/css`.
- Variables should be named using camelCase as commonly used in Grav.
- The use of config variables for css classes seem a bit overkill. The chances the plugin will be used in a project that already contains existing classes for summary/details etc is probably nihil. Using fixed class names might just be enough.
- Create an icon folder to store a default svg and load the svg from file. This allows users to use their own icon. Or, you could allow a config variable to contain an svg definition and load that definition.
- Variable `$matches` is only used in one `preg_match`

Hope this helps...

**Comment by bitstarr** [comment]
Thank you for your effort. Your recommendations are very helpful. I will overhaul the plugin asap.

**Comment by pamtbaau** [comment]
I need to correct a few things.

I said:
> The events `onMarkdownInitialized` and `onTwigSiteVariables` are being handled on each and every request, also when derived from Admin. Might be a good idea to move the event registrations into onPluginsInitialized and enable them only when request is not from Admin.

That's not entirely correct. Only `onTwigSiteVariables` is called on every page request. `onMarkdownInitialized` is only called when the page has not yet been cached, or when cache has been invalidated.

I also said:
> There is a dedicated event `onAssetsInitialized` where assets can be added.
> onAssetsInitialized
> The event indicates the assets manager has been initialized and is ready for assets to be added and managed.

That event only notifies the plugin that the Asset manager is ready to receive assets. It doesn't mean assets must be added during that event. If the plugin doesn't know during `onAssetsInitialized` if assets should be loaded, the assets can be loaded later.

OK, some remarks on this PR:
- I haven't looked into it very much and have not preference, but what is the reason for switching from Markdown extension to regex using a Twig template? You also might opt for using a shortcode extension, which makes parsing very easy. See for example: https://github.com/getgrav/grav-plugin-shortcode-core#as-a-custom-plugin
- Assets are loaded for every page request (from front-end) while not needed when page does not contain 'details'. 
Suggestion:
   - Remove registration of event `onAssetsInitialized`
   - Rename function `onAssetsInitialized` to `loadAssets`
   - During `onPageContentRaw`, when matches have been found and replaced, call at the end of the function `$this->loadAssets`. 
- You are overriding the default `config` variable in Twig. See https://learn.getgrav.org/17/themes/theme-vars#config-object.
You can access the config of the plugin using `{{ config.plugins['markdown-details'].content_class }}`
- The icon seems difficult to override now. Why not add an `icon` config and use that value in Twig template?
  ```
  // markdown-details.yaml
  icon: '<svg ...</svg>'
  ```
  ```
  // Twig
  {{ titleText }}<span>{{ config.plugins['markdown-details'].icon | raw }}
  ```
- Just a consideration. Use hardcoded unique class names (BEM or no BEM) in Twig and drop the extra config settings.
Personally I don't like BEM by the way. That's where SASS/SCSS comes in...
  ```
  .details {
    .title { }
    .trigger { }
  }
  ```

**Comment by bitstarr** [comment]
> what is the reason for switching from Markdown extension to regex using a Twig template?

I Think the markdown extension is not so easy to handle. Using regex to split the parts is more comprehensible. Also using a dedicated template was a request from a user which i can relate to, especially for customizing.

> You also might opt for using a shortcode extension

I wrote this plugin to only serve this one component in first place, because I did not needed all the shortcode functions in that specific project. It was in one of my first grav projects and I fell in love with grav at that time, so i wanted to give a little back.

> Assets are loaded for every page request (from front-end) while not needed when page does not contain 'details'. …

This is excellent, thanks for pointing this out.

> You are overriding the default config variable in Twig.

You are right. I wasn't aware of this.

> The icon seems difficult to override now. Why not add an `icon` config and use that value in Twig template? 

I find this part a bit tricky, because there are so many solutions to this. I opted for an easy one, where the customizing person can do some magic with a custom CSS or use the full potential of customizing the partial. The latter option also refers to your last point about CSS naming. Sometimes it is enough to set some new classes to fit the component in your custom theme without the need of overwriting the whole partial - so it's a neat extra. 

Thank you very much for taking time to review my tiny project, I really appreciate it!
